### PR TITLE
[AVG POOL] Asymmetric padding (SAME) support.

### PR DESCRIPTION
**Pull request by srkreddy1238**
Ref. https://discuss.tvm.ai/t/avgpool-same-padding/314

### CODE REVIEW

**Comment by srkreddy1238** [comment]
@tqchen  & @Huyuwei  Please have a look at this patch to support asymmetric padding with count_include_pad option.
padding attribute extended to accept 4 values by keeping 2 value logic intact.

Now the SAME padding option work perfectly for all operations.
Convolution : inserting pad operator before with pad_value = 0.
Maxpool : inserting pad operator before with pad_value=-inf
Avgpool : padding accepts (pad_top, pad_left, pad_bottom, pad_right)

**Comment by FrozenGene** [comment]
Do you want to continue to support Convolution / Maxpool and so on for SAME padding to accept 4 values? I have implemented it in my codebase, but my company has its own open source rule and need long time to contribute. So if you want to implement continually,   I am glad to see it. And you also need  to update WORKLOADS. i.e.:
```python
_Workload = namedtuple('Workload',
                       ['in_dtype', 'out_dtype', 'height', 'width', 'channel', 'multiplier',
                        'hkernel', 'wkernel', 'hpad', 'wpad', 'hstride', 'wstride'])
```
Here, it only has hpad / wpad，you should update to 4 elements. It also affects schedule. For example:
```python
    HPAD, WPAD = wkl.hpad, wkl.wpad  # You should update
    HSTR, WSTR = wkl.hstride, wkl.wstride

    VH, VW = sch.vh, sch.vw
    BC = sch.bc
    VC = sch.vc

   # You should update
    TH = H + 2*HPAD 
    TW = W + 2*WPAD
```

If you would like to do it, I would like to do code review it. Because I have implemented and work perfectly.

**Comment by srkreddy1238** [comment]
@FrozenGene I am glad you brought up this change in schedule. 
I will update the PR with schedule changes soon.

**Comment by srkreddy1238** [comment]
@FrozenGene  BTW, Convolution and Maxpool need not require these changes as there is alternative to achieve the same.

**Comment by FrozenGene** [comment]
You mean insert the pad operator? I think that is not an elegant solution. As @Huyuwei comment in this PR : https://github.com/dmlc/nnvm/pull/472 this pad operator brings overhead. Extend it will make us convert much easier and clearer.

**Comment by srkreddy1238** [comment]
I am open to extend the padding internal to Conv and Maxpool as well.

In any case there is no extra_pad in SAME option. We either insert pad at frontend (pad values will be [0,0] for actual operation) or we pass the pad values to the operation and the compute will add pad operation accordingly.

This PR is just accommodate count_include_pad effect with asymmetric padding.

**Comment by srkreddy1238** [comment]
@Huyuwei  & @tqchen any comment here ?

**Comment by srkreddy1238** [comment]
@Huyuwei  can you have a look at this ?

**Comment by Huyuwei** [comment]
Here is my suggestion:

- On topi implementation level, alwasys pass 4 pad values. It can be in the format of [height, left, bottom, right] or [[height, bottom], [left, right]]
- On nnvm symbol level, support api like this: https://keras.io/layers/convolutional/#zeropadding2d
- Support asymmetric padding for both avg and max pooling, to keep these two kinds of pooling consistent.

File: nnvm/include/nnvm/top/nn.h [hunk 1]
```diff
@@ -239,7 +239,10 @@ struct MaxPool2DParam : public dmlc::Parameter<MaxPool2DParam> {
       .describe("Specifies the strides of the convolution.");
     DMLC_DECLARE_FIELD(padding).set_default(TShape({0, 0}))
       .describe("If padding is non-zero, then the input is implicitly zero-padded"
-                "on both sides for padding number of points");
+                "Padding support both symmetric and asymmetric as"
+                "one int : same padding used on all sides"
+                "two int : bottom, right will use same padding as top, left"
+                "four int : padding just as given");
```
Comment (Huyuwei) [inline nnvm/include/nnvm/top/nn.h]: four int : padding width in the order of top, left, bottom, right

File: nnvm/include/nnvm/top/nn.h [hunk 2]
```diff
@@ -266,7 +269,10 @@ struct AvgPool2DParam : public dmlc::Parameter<AvgPool2DParam> {
...
```
Comment (Huyuwei) [inline nnvm/include/nnvm/top/nn.h]: four int : padding width in the order of top, left, bottom, right

File: nnvm/src/top/nn/pooling.cc [hunk 3]
```diff
@@ -44,23 +45,38 @@ inline bool Pool2DInferShape(const nnvm::NodeAttrs& attrs,
...
   const auto widx = layout.indexof('W');
 
+  TShape pad = param.padding;
+  if (param.padding.ndim() == 1) {
+    pad[1] = pad[0];
+    pad[2] = pad[0];
+    pad[3] = pad[0];
+  } else if (param.padding.ndim() == 2) {
+    pad[0] *= 2;
+    pad[1] *= 2;
```
Comment (Huyuwei) [inline nnvm/src/top/nn/pooling.cc]: better to use pad_h pad_w instead of pad[0] pad[1]
```python
pad_h = pad[0] * 2
pad_w = pad[1] * 2
```

File: nnvm/src/top/nn/pooling.cc [hunk 4]
```diff
@@ -44,23 +45,38 @@ inline bool Pool2DInferShape(const nnvm::NodeAttrs& attrs,
   const auto hidx = layout.indexof('H');
   const auto widx = layout.indexof('W');
 
+  TShape pad = param.padding;
+  if (param.padding.ndim() == 1) {
+    pad[1] = pad[0];
+    pad[2] = pad[0];
+    pad[3] = pad[0];
```
Comment (Huyuwei) [inline nnvm/src/top/nn/pooling.cc]: the logic here seems not correct, it should be:
```python
pad_h = pad[0] * 2
pad_w = pad[0] * 2
```

File: nnvm/include/nnvm/top/nn.h (hunk 1 above)
Comment (Huyuwei) [inline nnvm/include/nnvm/top/nn.h]: four int : padding width in the order of top, left, bottom, right

File: nnvm/src/top/nn/pooling.cc (hunk 4 above)
Comment (srkreddy1238) [reply nnvm/src/top/nn/pooling.cc]: Good catch, thanks.

**Comment by srkreddy1238** [comment]
@Huyuwei review comments handled.

File: topi/include/topi/nn/pooling.h [hunk 5]
```diff
@@ -52,28 +52,25 @@ inline Tensor pool_impl(const Tensor& x,
   CHECK(x->shape.size() >= 2) << "Pooling input must >= 2-D (H, W)";
   CHECK_EQ(kernel_size.size(), 2) << "Pooling kernel_size must have 2 elements";
   CHECK_EQ(stride_size.size(), 2) << "Pooling stride_size must have 2 elements";
-  CHECK_EQ(padding_size.size(), 2) << "Pooling padding_size must have 2 elements";
+  CHECK((padding_size.size() == 4)) << "Pooling padding_size must have 4 elements";
```
Comment (Huyuwei) [inline topi/include/topi/nn/pooling.h]: use CHECK_EQ

File: nnvm/tests/python/frontend/tensorflow/test_forward.py [hunk 6]
```diff
@@ -108,32 +108,32 @@ def _test_pooling(input_shape, **kwargs):
...
-                 window_shape=[1, 1],
-                 padding='SAME',
-                 pooling_type='AVG',
-                 dilation_rate=[1, 1],
-                 strides=[1, 1])
+    for padding in ["SAME", "VALID"]:
+        for pooling_type in ["MAX", "AVG"]:
+            for input_shape in [[2, 9, 10, 2], [2, 10, 9, 2]]:
+                for window_shape in [[1, 1], [2, 1], [2, 3]]:
+                    if padding != "SAME":
```
Comment (Huyuwei) [inline nnvm/tests/python/frontend/tensorflow/test_forward.py]: the testing script is not clean. I think it's ok to test a few important cases using the old style.

**Comment by srkreddy1238** [comment]
@Huyuwei patched as per review. Please check.

**Comment by srkreddy1238** [comment]
@Huyuwei thanks for the review.
